File: Python/card_detector.py
"""
Card Detection Module
Detects ID cards in camera frames using contour detection and aspect ratio filtering.
Enhanced with multi-scale detection, improved edge detection, and temporal smoothing.
"""
import cv2
import numpy as np


class CardDetector:
    """Detects ID cards with standard credit/debit card dimensions."""
    
    # Standard ID card aspect ratio (85.60mm × 53.98mm ≈ 1.586)
    # Made more lenient for better detection
    ASPECT_RATIO_MIN = 1.3  # Was 1.4
    ASPECT_RATIO_MAX = 1.8  # Was 1.7
    VERTICAL_ASPECT_MIN = 0.55  # Was 0.6 (1/1.8)
    VERTICAL_ASPECT_MAX = 0.77  # Was 0.75 (1/1.3)
    
    def __init__(self, min_area_ratio=0.02, max_area_ratio=0.95):
        """
        Initialize card detector.
        
        Args:
            min_area_ratio: Minimum card area as ratio of frame area (default: 2%)
            max_area_ratio: Maximum card area as ratio of frame area (default: 95% for close range)
        """
        self.min_area_ratio = min_area_ratio
        self.max_area_ratio = max_area_ratio
        # Size retention for stable detection
        self.last_detected_area = None
        self.area_tolerance = 0.5  # 50% variance allowed for size matching (more lenient)
        # Temporal smoothing for corners
        self.last_detected_corners = None
        self.corner_smoothing_alpha = 0.8  # Higher weight for more stability
        # Multi-scale detection - prioritize full scale for close range
        self.scale_factors = [1.0, 0.85, 0.7, 0.5]  # More scales, prioritize full size
        # Detection retention
        self.consecutive_failures = 0
        self.max_consecutive_failures = 10  # Allow up to 10 frames without detection before resetting
        self.has_ever_detected = False  # Track if we've ever detected a card
    
    def detect_card(self, frame):
        """
        Detect ID card in the frame using multi-scale detection with improved retention.
        
        Args:
            frame: Input BGR frame from camera
        
        Returns:
            tuple: (card_found: bool, corners: np.array or None)
                corners is a 4x2 array of card corner coordinates if found
        """
        if frame is None:
            return False, None
        
        # Try multi-scale detection - prioritize full scale first for close range
        best_result = None
        best_score = 0
        
        for scale in self.scale_factors:
            if scale != 1.0:
                # Resize frame for multi-scale detection
                h, w = frame.shape[:2]
                scaled_frame = cv2.resize(frame, (int(w * scale), int(h * scale)))
            else:
                scaled_frame = frame
            
            result = self._detect_at_scale(scaled_frame, scale)
            
            if result is not None:
                corners, score = result
                # Scale corners back to original frame size
                if scale != 1.0:
                    corners = corners / scale
                
                # Boost score for full-scale detection (close range)
                if scale == 1.0:
                    score *= 1.3
                
                if score > best_score:
                    best_score = score
                    best_result = corners
        
        if best_result is not None:
            # Successful detection
            self.consecutive_failures = 0
            self.has_ever_detected = True
            
            # Apply temporal smoothing
            if self.last_detected_corners is not None:
                # Exponential smoothing of corners - more aggressive smoothing for stability
                smoothed_corners = (
                    self.corner_smoothing_alpha * self.last_detected_corners +
                    (1 - self.corner_smoothing_alpha) * best_result
                )
                self.last_detected_corners = smoothed_corners
                return True, smoothed_corners
            else:
                self.last_detected_corners = best_result
                return True, best_result
        
        # No detection in this frame
        self.consecutive_failures += 1
        
        # Retention logic: if we've detected before and failures are below threshold, retain last detection
        if self.has_ever_detected and self.consecutive_failures <= self.max_consecutive_failures:
            if self.last_detected_corners is not None:
                # Retain last known corners - don't lose detection immediately
                return True, self.last_detected_corners
        
        # Too many failures or never detected - reset tracking
        if self.consecutive_failures > self.max_consecutive_failures:
            self.last_detected_corners = None
            self.has_ever_detected = False
            self.consecutive_failures = 0
        
        return False, None
    
    def _detect_at_scale(self, frame, scale=1.0):
        """
        Detect card at a specific scale.
        
        Args:
            frame: Input frame at this scale
            scale: Scale factor (1.0 = full size, <1.0 = downscaled)
        
        Returns:
            tuple: (corners, score) or None if not found
        """
        # Convert to grayscale
        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        
        # Apply Gaussian blur to reduce noise
        blurred = cv2.GaussianBlur(gray, (5, 5), 0)
        
        # Enhanced edge detection with multiple methods
        # Adjust parameters based on scale (close range needs different handling)
        is_close_range = scale >= 1.0
        
        # Method 1: Multi-threshold Canny
        # For close range, use higher thresholds to reduce noise
        if is_close_range:
            edges1 = cv2.Canny(blurred, 30, 100)
            edges2 = cv2.Canny(blurred, 50, 150)
            edges3 = cv2.Canny(blurred, 40, 120)
        else:
            edges1 = cv2.Canny(blurred, 20, 80)
            edges2 = cv2.Canny(blurred, 40, 120)
            edges3 = cv2.Canny(blurred, 30, 100)
        
        edges = cv2.bitwise_or(edges1, edges2)
        edges = cv2.bitwise_or(edges, edges3)
        
        # Method 2: Adaptive thresholding for low contrast
        # Use larger block size for close range to reduce noise
        block_size = 15 if is_close_range else 11
        adaptive_thresh = cv2.adaptiveThreshold(
            blurred, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C, 
            cv2.THRESH_BINARY_INV, block_size, 2
        )
        edges = cv2.bitwise_or(edges, adaptive_thresh)
        
        # Method 3: Sobel edge detection
        sobel_x = cv2.Sobel(blurred, cv2.CV_64F, 1, 0, ksize=3)
        sobel_y = cv2.Sobel(blurred, cv2.CV_64F, 0, 1, ksize=3)
        sobel = np.sqrt(sobel_x**2 + sobel_y**2)
        if np.max(sobel) > 0:
            sobel = np.uint8(255 * sobel / np.max(sobel))
            sobel_thresh_val = 60 if is_close_range else 50
            _, sobel_thresh = cv2.threshold(sobel, sobel_thresh_val, 255, cv2.THRESH_BINARY)
            edges = cv2.bitwise_or(edges, sobel_thresh)
        
        # Morphological operations to close gaps
        # More aggressive for close range to handle larger gaps
        kernel_size = 5 if is_close_range else 3
        kernel = np.ones((kernel_size, kernel_size), np.uint8)
        iterations = 3 if is_close_range else 2
        edges = cv2.morphologyEx(edges, cv2.MORPH_CLOSE, kernel, iterations=iterations)
        edges = cv2.dilate(edges, kernel, iterations=iterations)
        edges = cv2.erode(edges, kernel, iterations=1)
        
        # Find contours
        contours, _ = cv2.findContours(edges, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        
        if not contours:
            return None
        
        # Filter contours by area and shape
        frame_area = frame.shape[0] * frame.shape[1]
        min_area = frame_area * self.min_area_ratio
        max_area = frame_area * self.max_area_ratio
        
        # For close range (full scale), be more lenient with area
        if scale >= 1.0:
            # Allow slightly larger areas for close-up detection
            max_area = frame_area * min(0.98, self.max_area_ratio * 1.05)
        
        valid_contours = []
        
        for contour in contours:
            area = cv2.contourArea(contour)
            
            # Check area bounds
            if area < min_area or area > max_area:
                continue
            
            # Check convexity defects for better shape validation
            hull = cv2.convexHull(contour)
            hull_area = cv2.contourArea(hull)
            solidity = 1.0  # Default solidity
            if hull_area > 0:
                solidity = area / hull_area
                # Reject shapes that are too irregular (solidity < 0.85)
                if solidity < 0.85:
                    continue
            
            # Approximate contour to polygon
            peri = cv2.arcLength(contour, True)
            if peri == 0:
                continue
            
            # Try different epsilon values to get 4 vertices (preferred) or 3-5 (acceptable)
            approx = None
            vertex_count = None
            for epsilon_factor in [0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08]:
                test_approx = cv2.approxPolyDP(contour, epsilon_factor * peri, True)
                v_count = len(test_approx)
                if v_count == 4:
                    # Perfect 4 vertices
                    approx = test_approx
                    vertex_count = 4
                    break
                elif v_count >= 3 and v_count <= 5 and approx is None:
                    # Accept 3-5 vertices as backup
                    approx = test_approx
                    vertex_count = v_count
            
            # Need at least 3 vertices to form a rectangle
            if approx is None or vertex_count < 3:
                continue
            
            # Get corners
            if vertex_count == 4:
                corners = approx.reshape(4, 2).astype(np.float32)
                # Validate perspective rectangle - more lenient for close range
                if not self._validate_perspective_rectangle(corners, scale):
                    continue
            else:
                # For 3 or 5 vertices, create a bounding rectangle
                x, y, w, h = cv2.boundingRect(approx)
                if w == 0 or h == 0:
                    continue
                corners = np.array([
                    [x, y],           # top-left
                    [x + w, y],       # top-right
                    [x + w, y + h],   # bottom-right
                    [x, y + h]        # bottom-left
                ], dtype=np.float32)
            
            # Calculate aspect ratio from corners
            x, y, w, h = cv2.boundingRect(corners.astype(np.int32))
            aspect_ratio = float(w) / h if h > 0 else 0
            inverse_aspect = float(h) / w if w > 0 else 0
            
            # Check if aspect ratio matches ID card dimensions
            is_horizontal = (self.ASPECT_RATIO_MIN <= aspect_ratio <= self.ASPECT_RATIO_MAX)
            is_vertical = (self.VERTICAL_ASPECT_MIN <= inverse_aspect <= self.VERTICAL_ASPECT_MAX)
            
            if is_horizontal or is_vertical:
                # Calculate score based on multiple factors
                score = self._calculate_detection_score(corners, area, solidity)
                valid_contours.append((corners, area, score))
        
        if not valid_contours:
            return None
        
        # Sort by score (higher is better)
        valid_contours.sort(key=lambda x: x[2], reverse=True)
        
        # If we have a previous detection, prefer similar-sized cards
        # More aggressive matching for retention
        if self.last_detected_area is not None:
            # Boost score for similar size - more lenient matching
            for i, (corners, area, score) in enumerate(valid_contours):
                area_diff = abs(area - self.last_detected_area) / self.last_detected_area
                if area_diff <= self.area_tolerance:
                    # Strong boost for similar size (helps retention)
                    valid_contours[i] = (corners, area, score * 2.0)
                elif area_diff <= self.area_tolerance * 1.5:
                    # Moderate boost for somewhat similar size
                    valid_contours[i] = (corners, area, score * 1.2)
            valid_contours.sort(key=lambda x: x[2], reverse=True)
        
        best_corners, best_area, _ = valid_contours[0]
        
        # Update last detected area - more stable updates for retention
        if self.last_detected_area is None:
            self.last_detected_area = best_area
        else:
            # Exponential moving average - slower update for stability
            # This helps retain detection even when size changes slightly
            self.last_detected_area = 0.8 * self.last_detected_area + 0.2 * best_area
        
        # Order corners: top-left, top-right, bottom-right, bottom-left
        ordered_corners = self._order_corners(best_corners)
        
        return (ordered_corners, valid_contours[0][2])
    
    def _validate_perspective_rectangle(self, corners, scale=1.0):
        """
        Validate that corners form a valid perspective rectangle.
        More lenient for close range (full scale) detection.
        
        Args:
            corners: 4x2 array of corner coordinates
            scale: Scale factor (more lenient for scale >= 1.0)
        
        Returns:
            bool: True if valid perspective rectangle
        """
        if corners.shape[0] != 4:
            return False
        
        # More lenient thresholds for close range
        is_close_range = scale >= 1.0
        parallel_threshold = 0.6 if is_close_range else 0.7
        angle_threshold = 0.4 if is_close_range else 0.3
        
        # Check that opposite sides are roughly parallel
        # Calculate vectors for each side
        v1 = corners[1] - corners[0]  # Top edge
        v2 = corners[2] - corners[1]   # Right edge
        v3 = corners[3] - corners[2]  # Bottom edge
        v4 = corners[0] - corners[3]  # Left edge
        
        # Check for zero-length edges
        if np.linalg.norm(v1) < 1 or np.linalg.norm(v2) < 1 or \
           np.linalg.norm(v3) < 1 or np.linalg.norm(v4) < 1:
            return False
        
        # Normalize vectors
        def normalize(v):
            norm = np.linalg.norm(v)
            return v / norm if norm > 0 else v
        
        v1_norm = normalize(v1)
        v3_norm = normalize(v3)
        v2_norm = normalize(v2)
        v4_norm = normalize(v4)
        
        # Check parallelism (dot product should be close to 1 or -1)
        top_bottom_parallel = abs(np.dot(v1_norm, v3_norm)) > parallel_threshold
        left_right_parallel = abs(np.dot(v2_norm, v4_norm)) > parallel_threshold
        
        # Check that angles are roughly 90 degrees (more lenient for close range)
        top_right_angle = abs(np.dot(v1_norm, v2_norm)) < angle_threshold
        
        # For close range, only require 2 out of 3 checks to pass
        if is_close_range:
            checks_passed = sum([top_bottom_parallel, left_right_parallel, top_right_angle])
            return checks_passed >= 2
        
        return top_bottom_parallel and left_right_parallel and top_right_angle
    
    def _calculate_detection_score(self, corners, area, solidity):
        """
        Calculate detection score based on multiple factors.
        
        Args:
            corners: Detected corners
            area: Contour area
            solidity: Contour solidity (area/hull_area)
        
        Returns:
            float: Detection score
        """
        score = 1.0
        
        # Prefer larger areas (up to a point)
        score *= min(area / 10000, 2.0)  # Normalize
        
        # Prefer higher solidity
        score *= solidity
        
        # Prefer rectangles with 4 vertices
        if len(corners) == 4:
            score *= 1.2
        
        return score
    
    def reset_size_tracking(self):
        """Reset size tracking (call when card is no longer visible)."""
        self.last_detected_area = None
        self.last_detected_corners = None
        self.consecutive_failures = 0
        self.has_ever_detected = False
    
    def _order_corners(self, corners):
        """
        Order corners in consistent order: top-left, top-right, bottom-right, bottom-left.
        
        Args:
            corners: 4x2 array of corner coordinates
        
        Returns:
            4x2 array of ordered corners
        """
        # Reshape to (4, 2)
        corners = corners.reshape(4, 2)
        
        # Calculate sum and difference
        sum_points = corners.sum(axis=1)
        diff_points = np.diff(corners, axis=1)
        
        # Top-left: smallest sum
        # Bottom-right: largest sum
        # Top-right: smallest difference
        # Bottom-left: largest difference
        top_left = corners[np.argmin(sum_points)]
        bottom_right = corners[np.argmax(sum_points)]
        top_right = corners[np.argmin(diff_points)]
        bottom_left = corners[np.argmax(diff_points)]
        
        return np.array([top_left, top_right, bottom_right, bottom_left], dtype=np.float32)

File: Flutter/lib/services/image_processing/card_detector.dart
import 'dart:typed_data';
import 'dart:math' as math;
import 'package:image/image.dart' as img;

/// Card Detection Module
/// Detects ID cards in camera frames using edge detection and contour analysis.
class CardDetector {
  // Standard ID card aspect ratio (85.60mm × 53.98mm ≈ 1.586)
  static const double aspectRatioMin = 1.3;
  static const double aspectRatioMax = 1.8;
  static const double verticalAspectMin = 0.55; // 1/1.8
  static const double verticalAspectMax = 0.77; // 1/1.3

  final double minAreaRatio;
  final double maxAreaRatio;
  
  // Size retention for stable detection
  double? _lastDetectedArea;
  final double areaTolerance = 0.3; // 30% variance allowed

  CardDetector({
    this.minAreaRatio = 0.02,
    this.maxAreaRatio = 0.85,
  });

  /// Detect ID card in the frame
  /// Returns: (cardFound: bool, corners: List<Point>? or null)
  /// corners is a list of 4 points [top-left, top-right, bottom-right, bottom-left]
  Future<({bool found, List<math.Point<int>>? corners})> detectCard(
    Uint8List imageBytes,
  ) async {
    try {
      // Decode image
      final image = img.decodeImage(imageBytes);
      if (image == null) {
        return (found: false, corners: null);
      }

      // Convert to grayscale
      final gray = img.grayscale(image);

      // Apply Gaussian blur
      final blurred = img.gaussianBlur(gray, radius: 5);

      // Edge detection using Canny algorithm
      final edges = _cannyEdgeDetection(blurred);

      // Find contours
      final contours = _findContours(edges);

      if (contours.isEmpty) {
        return (found: false, corners: null);
      }

      // Filter contours by area and shape
      final frameArea = image.width * image.height;
      final minArea = frameArea * minAreaRatio;
      final maxArea = frameArea * maxAreaRatio;

      final validContours = <({List<math.Point<int>> corners, double area})>[];

      for (final contour in contours) {
        final area = _calculatePolygonArea(contour);
        
        // Check area bounds
        if (area < minArea || area > maxArea) {
          continue;
        }

        // Approximate polygon
        final approx = _approximatePolygon(contour);
        
        if (approx.length < 3) {
          continue;
        }

        // Get bounding rectangle
        final bounds = _getBoundingRect(approx);
        final width = bounds.width;
        final height = bounds.height;

        if (width == 0 || height == 0) {
          continue;
        }

        // Calculate aspect ratio
        final aspectRatio = width / height;
        final inverseAspect = height / width;

        // Check if aspect ratio matches ID card dimensions
        final isHorizontal = aspectRatio >= aspectRatioMin && aspectRatio <= aspectRatioMax;
        final isVertical = inverseAspect >= verticalAspectMin && inverseAspect <= verticalAspectMax;

        if (isHorizontal || isVertical) {
          // Create corners from bounding rect or use approximated polygon
          List<math.Point<int>> corners;
          if (approx.length == 4) {
            corners = _orderCorners(approx);
          } else {
            // Create bounding rectangle corners
            corners = [
              math.Point(bounds.left, bounds.top),
              math.Point(bounds.right, bounds.top),
              math.Point(bounds.right, bounds.bottom),
              math.Point(bounds.left, bounds.bottom),
            ];
          }
          validContours.add((corners: corners, area: area));
        }
      }

      if (validContours.isEmpty) {
        _resetSizeTracking();
        return (found: false, corners: null);
      }

      // Select best contour (prefer similar size if we have previous detection)
      final best = _selectBestContour(validContours);
      
      // Update last detected area
      _lastDetectedArea = best.area;

      return (found: true, corners: best.corners);
    } catch (e) {
      // Error handled silently
      return (found: false, corners: null);
    }
  }

  /// Reset size tracking (call when card is no longer visible)
  void resetSizeTracking() {
    _lastDetectedArea = null;
  }

  void _resetSizeTracking() {
    _lastDetectedArea = null;
  }

  /// Canny edge detection implementation
  Uint8List _cannyEdgeDetection(img.Image gray) {
    // Apply Sobel operator for gradient
    final sobelX = _sobelX(gray);
    final sobelY = _sobelY(gray);
    
    // Calculate gradient magnitude and direction
    final gradient = _calculateGradient(sobelX, sobelY);
    
    // Non-maximum suppression
    final suppressed = _nonMaximumSuppression(gradient);
    
    // Double threshold and edge tracking
    return _doubleThreshold(suppressed, lowThreshold: 20, highThreshold: 80);
  }

  Uint8List _sobelX(img.Image gray) {
    final result = Uint8List(gray.width * gray.height);
    final kernel = [-1, 0, 1, -2, 0, 2, -1, 0, 1];
    
    for (var y = 1; y < gray.height - 1; y++) {
      for (var x = 1; x < gray.width - 1; x++) {
        var sum = 0.0;
        for (var ky = 0; ky < 3; ky++) {
          for (var kx = 0; kx < 3; kx++) {
            final px = gray.getPixel(x + kx - 1, y + ky - 1);
            sum += (px.r * kernel[ky * 3 + kx]);
          }
        }
        result[y * gray.width + x] = sum.abs().round().clamp(0, 255);
      }
    }
    return result;
  }

  Uint8List _sobelY(img.Image gray) {
    final result = Uint8List(gray.width * gray.height);
    final kernel = [-1, -2, -1, 0, 0, 0, 1, 2, 1];
    
    for (var y = 1; y < gray.height - 1; y++) {
      for (var x = 1; x < gray.width - 1; x++) {
        var sum = 0.0;
        for (var ky = 0; ky < 3; ky++) {
          for (var kx = 0; kx < 3; kx++) {
            final px = gray.getPixel(x + kx - 1, y + ky - 1);
            sum += (px.r * kernel[ky * 3 + kx]);
          }
        }
        result[y * gray.width + x] = sum.abs().round().clamp(0, 255);
      }
    }
    return result;
  }

  List<({int magnitude, double direction})> _calculateGradient(
    Uint8List sobelX,
    Uint8List sobelY,
  ) {
    final gradient = <({int magnitude, double direction})>[];
    for (var i = 0; i < sobelX.length; i++) {
      final gx = sobelX[i].toDouble();
      final gy = sobelY[i].toDouble();
      final magnitude = math.sqrt(gx * gx + gy * gy).toInt();
      final direction = math.atan2(gy, gx);
      gradient.add((magnitude: magnitude, direction: direction));
    }
    return gradient;
  }

  Uint8List _nonMaximumSuppression(
    List<({int magnitude, double direction})> gradient,
  ) {
    final width = math.sqrt(gradient.length).toInt();
    final height = width;
    final result = Uint8List(gradient.length);
    
    for (var y = 1; y < height - 1; y++) {
      for (var x = 1; x < width - 1; x++) {
        final idx = y * width + x;
        final mag = gradient[idx].magnitude;
        final dir = gradient[idx].direction;
        
        // Determine neighbors based on gradient direction
        int neighbor1, neighbor2;
        if ((dir >= -math.pi / 8 && dir < math.pi / 8) ||
            (dir >= 7 * math.pi / 8 || dir < -7 * math.pi / 8)) {
          // Horizontal
          neighbor1 = gradient[idx - 1].magnitude;
          neighbor2 = gradient[idx + 1].magnitude;
        } else if ((dir >= math.pi / 8 && dir < 3 * math.pi / 8) ||
                   (dir >= -7 * math.pi / 8 && dir < -5 * math.pi / 8)) {
          // Diagonal 1
          neighbor1 = gradient[(y - 1) * width + (x + 1)].magnitude;
          neighbor2 = gradient[(y + 1) * width + (x - 1)].magnitude;
        } else if ((dir >= 3 * math.pi / 8 && dir < 5 * math.pi / 8) ||
                   (dir >= -5 * math.pi / 8 && dir < -3 * math.pi / 8)) {
          // Vertical
          neighbor1 = gradient[(y - 1) * width + x].magnitude;
          neighbor2 = gradient[(y + 1) * width + x].magnitude;
        } else {
          // Diagonal 2
          neighbor1 = gradient[(y - 1) * width + (x - 1)].magnitude;
          neighbor2 = gradient[(y + 1) * width + (x + 1)].magnitude;
        }
        
        if (mag >= neighbor1 && mag >= neighbor2) {
          result[idx] = mag.clamp(0, 255).toInt();
        } else {
          result[idx] = 0;
        }
      }
    }
    return result;
  }

  Uint8List _doubleThreshold(
    Uint8List suppressed, {
    required int lowThreshold,
    required int highThreshold,
  }) {
    final result = Uint8List(suppressed.length);
    final width = math.sqrt(suppressed.length).toInt();
    final height = width;
    
    // Mark strong and weak edges
    for (var i = 0; i < suppressed.length; i++) {
      if (suppressed[i] >= highThreshold) {
        result[i] = 255; // Strong edge
      } else if (suppressed[i] >= lowThreshold) {
        result[i] = 128; // Weak edge
      } else {
        result[i] = 0;
      }
    }
    
    // Edge tracking by hysteresis
    for (var y = 1; y < height - 1; y++) {
      for (var x = 1; x < width - 1; x++) {
        final idx = y * width + x;
        if (result[idx] == 128) {
          // Check if connected to strong edge
          bool connected = false;
          for (var dy = -1; dy <= 1; dy++) {
            for (var dx = -1; dx <= 1; dx++) {
              if (dx == 0 && dy == 0) continue;
              final nIdx = (y + dy) * width + (x + dx);
              if (result[nIdx] == 255) {
                connected = true;
                break;
              }
            }
            if (connected) break;
          }
          result[idx] = connected ? 255 : 0;
        }
      }
    }
    
    return result;
  }

  /// Find contours in binary image
  List<List<math.Point<int>>> _findContours(Uint8List binary) {
    final width = math.sqrt(binary.length).toInt();
    final height = width;
    final visited = List.generate(width * height, (_) => false);
    final contours = <List<math.Point<int>>>[];
    
    for (var y = 0; y < height; y++) {
      for (var x = 0; x < width; x++) {
        final idx = y * width + x;
        if (binary[idx] == 255 && !visited[idx]) {
          // Found new contour
          final contour = <math.Point<int>>[];
          _traceContour(binary, visited, width, height, x, y, contour);
          if (contour.length >= 4) {
            contours.add(contour);
          }
        }
      }
    }
    
    return contours;
  }

  void _traceContour(
    Uint8List binary,
    List<bool> visited,
    int width,
    int height,
    int startX,
    int startY,
    List<math.Point<int>> contour,
  ) {
    final stack = <math.Point<int>>[math.Point(startX, startY)];
    
    while (stack.isNotEmpty) {
      final point = stack.removeLast();
      final x = point.x;
      final y = point.y;
      
      if (x < 0 || x >= width || y < 0 || y >= height) continue;
      final idx = y * width + x;
      if (visited[idx] || binary[idx] != 255) continue;
      
      visited[idx] = true;
      contour.add(point);
      
      // Add 8-connected neighbors
      for (var dy = -1; dy <= 1; dy++) {
        for (var dx = -1; dx <= 1; dx++) {
          if (dx == 0 && dy == 0) continue;
          stack.add(math.Point(x + dx, y + dy));
        }
      }
    }
  }

  /// Approximate polygon using Douglas-Peucker algorithm
  List<math.Point<int>> _approximatePolygon(
    List<math.Point<int>> contour, {
    double epsilon = 0.02,
  }) {
    if (contour.length <= 2) return contour;
    
    final peri = _calculatePerimeter(contour);
    final epsilonScaled = epsilon * peri;
    
    return _douglasPeucker(contour, epsilonScaled);
  }

  List<math.Point<int>> _douglasPeucker(
    List<math.Point<int>> points,
    double epsilon,
  ) {
    if (points.length <= 2) return points;
    
    // Find point with maximum distance from line
    double maxDist = 0;
    int maxIndex = 0;
    final start = points.first;
    final end = points.last;
    
    for (var i = 1; i < points.length - 1; i++) {
      final dist = _pointToLineDistance(points[i], start, end);
      if (dist > maxDist) {
        maxDist = dist;
        maxIndex = i;
      }
    }
    
    if (maxDist > epsilon) {
      // Recursively simplify
      final left = _douglasPeucker(points.sublist(0, maxIndex + 1), epsilon);
      final right = _douglasPeucker(points.sublist(maxIndex), epsilon);
      
      // Combine (remove duplicate point)
      return [...left.sublist(0, left.length - 1), ...right];
    } else {
      // Return only endpoints
      return [start, end];
    }
  }

  double _pointToLineDistance(
    math.Point<int> point,
    math.Point<int> lineStart,
    math.Point<int> lineEnd,
  ) {
    final dx = lineEnd.x - lineStart.x;
    final dy = lineEnd.y - lineStart.y;
    
    if (dx == 0 && dy == 0) {
      return math.sqrt(
        math.pow(point.x - lineStart.x, 2) + math.pow(point.y - lineStart.y, 2),
      );
    }
    
    final t = ((point.x - lineStart.x) * dx + (point.y - lineStart.y) * dy) /
        (dx * dx + dy * dy);
    
    final clampedT = t.clamp(0.0, 1.0);
    final projX = lineStart.x + clampedT * dx;
    final projY = lineStart.y + clampedT * dy;
    
    return math.sqrt(
      math.pow(point.x - projX, 2) + math.pow(point.y - projY, 2),
    );
  }

  double _calculatePerimeter(List<math.Point<int>> points) {
    double peri = 0;
    for (var i = 0; i < points.length; i++) {
      final next = points[(i + 1) % points.length];
      final dx = next.x - points[i].x;
      final dy = next.y - points[i].y;
      peri += math.sqrt(dx * dx + dy * dy);
    }
    return peri;
  }

  double _calculatePolygonArea(List<math.Point<int>> points) {
    // Shoelace formula
    double area = 0;
    for (var i = 0; i < points.length; i++) {
      final j = (i + 1) % points.length;
      area += points[i].x * points[j].y;
      area -= points[j].x * points[i].y;
    }
    return area.abs() / 2.0;
  }

  ({int left, int top, int right, int bottom, int width, int height})
      _getBoundingRect(List<math.Point<int>> points) {
    int minX = points[0].x;
    int maxX = points[0].x;
    int minY = points[0].y;
    int maxY = points[0].y;
    
    for (final point in points) {
      minX = math.min(minX, point.x);
      maxX = math.max(maxX, point.x);
      minY = math.min(minY, point.y);
      maxY = math.max(maxY, point.y);
    }
    
    return (
      left: minX,
      top: minY,
      right: maxX,
      bottom: maxY,
      width: maxX - minX,
      height: maxY - minY,
    );
  }

  List<math.Point<int>> _orderCorners(List<math.Point<int>> corners) {
    // Order corners: top-left, top-right, bottom-right, bottom-left
    // Calculate sum and difference
    final sums = corners.map((p) => p.x + p.y).toList();
    final diffs = corners.map((p) => p.x - p.y).toList();
    
    final topLeft = corners[sums.indexOf(sums.reduce(math.min))];
    final bottomRight = corners[sums.indexOf(sums.reduce(math.max))];
    final topRight = corners[diffs.indexOf(diffs.reduce(math.min))];
    final bottomLeft = corners[diffs.indexOf(diffs.reduce(math.max))];
    
    return [topLeft, topRight, bottomRight, bottomLeft];
  }

  ({List<math.Point<int>> corners, double area}) _selectBestContour(
    List<({List<math.Point<int>> corners, double area})> contours,
  ) {
    if (_lastDetectedArea == null) {
      // Sort by area (largest first)
      contours.sort((a, b) => b.area.compareTo(a.area));
      return contours.first;
    }
    
    // Prefer similar size
    contours.sort((a, b) {
      final aDiff = ((a.area - _lastDetectedArea!).abs() / _lastDetectedArea!);
      final bDiff = ((b.area - _lastDetectedArea!).abs() / _lastDetectedArea!);
      
      if (aDiff <= areaTolerance && bDiff <= areaTolerance) {
        return b.area.compareTo(a.area); // Prefer larger
      } else if (aDiff <= areaTolerance) {
        return -1; // a is better
      } else if (bDiff <= areaTolerance) {
        return 1; // b is better
      } else {
        return aDiff.compareTo(bDiff); // Prefer closer
      }
    });
    
    return contours.first;
  }
}

